### DISCUSSION TITLE
Ideas how to make a "StickyHeaderList" with a `JList` and or `JTable`

### DISCUSSION THREAD

**Discussion by bric3** [aterai/java-swing-tips#28] [Q&A] [answered]
I saw this pattern used quite a lot in mobile apps, and I think this could really be interesting for desktop as well.

![DoubleStickyHeaderList](https://github.com/user-attachments/assets/63a6d42a-81ea-487e-a8ef-6597612345f7)

**Comment by aterai** [comment]
It looks like Sticky Header can be implemented in Swing by using `JLayer`. If possible, I would like to create a example in the next month or so.

**Reply by bric3** [reply]
Ah interesting! I didn't think about that, I was thinking about making something inspired by the materialized row of IntelliJ [`JListHoveredRowMaterialiser`](https://github.com/JetBrains/intellij-community/blob/29c3dfca926b94b5f3a8239468ad25ef06e47588/platform/collaboration-tools/src/com/intellij/collaboration/ui/util/JListHoveredRowMaterialiser.kt), but the `JLayer` might be a better approach here !

**Reply by aterai** [reply, marked as answer]
Before creating multiple sticky headers, we created an example of a simple single sticky header JList. https://github.com/aterai/java-swing-tips/pull/29
Please check and modify the operation when you have time.

**Reply by bric3** [reply]
Hi, thank you for looking into it !
Note I was mostly mentioning the single sticky header _UI pattern_. Not necessarily looking into the "multiple" one.

By the way, I noticed the scrolling speed is different between the two lists. The one with the Layer seems slower (using the macbook trackpad), I'm not yet sure why.

**Reply by bric3** [reply]
Profiling the code shows indeed more time spend in the `StickyLayerUI.paint`. Not sure yet what to optimize, the `super.paint` looks like the one taking much of the time. Also more time is spent delivering the mouse events with the sticky header list, but this could be a side effect 🤔.

**with stick header**
![image](https://github.com/user-attachments/assets/6b969971-c47e-4c91-bdc6-da4724eda5dc)

**regular list scrolling**
![image](https://github.com/user-attachments/assets/db7821c3-a00f-43f6-adeb-39508292c0d6)

**JFR profiles**
[2024-11-25T1302-StickyHeaderList-profiles.zip](https://github.com/user-attachments/files/17902738/2024-11-25T1302-StickyHeaderList-profiles.zip)

**Reply by aterai** [reply]
Thanks for profiling the code😀.

> the scrolling speed is different between the two lists

In the default `JList` on the left, the mouse wheel scrolls in cell blocks, but in the `JList` on the right, you can set `JList#getVerticalScrollBar()#setUnitIncrement(2);` to scroll in pixels to It is easier to see the sticky header replacement. The scroll amount of 2px seems to be too slow, so modified it to 4px in both `JList`s.

> Profiling the code shows indeed more time spend in the StickyLayerUI.paint

Since there is no need to redraw when moving the mouse, have modified it to exclude this.

**Reply by bric3** [reply]
@aterai Ah, I missed the unit increment indeed! This clearly has an effect on scrolling speed.

I pushed a bit the curiosity and I experimented using various increment values, and discovered that the JDK's Aqua ScrollBar contributes ~7% of the samples for almost any value of the unit increment, and the other contribution is usually the list UI that takes around 4%.

<img width="691" alt="image" src="https://github.com/user-attachments/assets/bb78ee39-6596-4d43-a03c-224c9d32b997">

![image](https://github.com/user-attachments/assets/e8eef5af-2d13-428a-8802-08fe2938c0b7)

![image](https://github.com/user-attachments/assets/dc7d6a2c-bb11-4e20-95a6-c3b05e936df4)

And even without setting the unit increment (i.e. it's per cell blocks), the `AquaScrollBarUI.paint` takes a significant time 7% - 10%.

![image](https://github.com/user-attachments/assets/c795c9ac-aacf-45f0-830b-ec691264ff57)

I wonder if this is something more visible on HiDPI screens.
----

For reference I'm running corretto-23.0.1.8.1 on macOs 15.1.1.

----

By the way I used IntelliJ Ultimate that integrates Async Profiler under the hood. And allows to visualize the JFR recording within IJ. I don't know if it's available on Windows though.

**Reply by Atera** [reply]
> @Atera Ah, I missed the unit increment indeed! This clearly has an effect on scrolling speed.

I think you've summoned the wrong person :)

**Reply by bric3** [reply]
Sorry @Atera :)
Fixed the typo in the handle in my message above.
